### PR TITLE
newlib: Bump to 4.3.0.20230120

### PR DESCRIPTION
**Pull request by alistair23**
A follow up of https://github.com/tock/libtock-c/pull/329

### CODE REVIEW

**Comment by bradjc** [comment]
Why are the compiled archives so much smaller?

File: libc++/Makefile [hunk 1]
```diff
@@ -16,9 +16,9 @@ rebuild-gcc: gcc-$(GCC_VERSION)
 	@mkdir -p gcc-$(GCC_VERSION)-out
 	@echo "Entering directory gcc-$(GCC_VERSION)-out"
 	cd gcc-$(GCC_VERSION)-out; sh ../build.sh ../$<
-	cp gcc-$(GCC_VERSION)-out/arm-none-eabi/libgcc/libgcc.a .
-	cp gcc-$(GCC_VERSION)-out/arm-none-eabi/libstdc++-v3/src/.libs/libstdc++.a .
-	cp gcc-$(GCC_VERSION)-out/arm-none-eabi/libstdc++-v3/libsupc++/.libs/libsupc++.a .
+	cp gcc-$(GCC_VERSION)-out/arm-none-eabi/libgcc/libgcc.a ./cortex-m/
+	cp gcc-$(GCC_VERSION)-out/arm-none-eabi/libstdc++-v3/src/.libs/libstdc++.a ./cortex-m/
+	cp gcc-$(GCC_VERSION)-out/arm-none-eabi/libstdc++-v3/libsupc++/.libs/libsupc++.a ./cortex-m/
```
Comment (bradjc) [inline libc++/Makefile:21]: Do these folders exist?

**Comment by alistair23** [comment]
I'm not sure why. Maybe it's built with a newer compiler?

**Comment by hudson-ayers** [comment]
Why are the C++ compiled archives removed?

**Comment by bradjc** [comment]
I think we are wondering about the same thing. Where is the cortex-m folder that should have the c++ libraries?

**Comment by alistair23** [comment]
Good catch. I didn't check them and didn't notice as the CI passed.

I have rebulit them and added them now.

It would be good for others to test this as well. Everything seems to work fine for me, but others might have issues with different compilers

**Comment by bradjc** [comment]
I think this has come up before, somewhere, but we inadvertently exist where we compile our own newlib but then also require newlib to be installed on the user's machine. The reason is we need the .h files. I'm running into this because homebrew upstreamed the arm-none-eabi-gcc formula but not newlib.

I'd like to add committing the include directory `./newlib-4*/newlib/libc/include/` with this PR so we can remove the need for users to have newlib installed.

**Comment by bradjc** [comment]
```diff
diff --git a/Configuration.mk b/Configuration.mk
index 166cbd1..0021613 100644
--- a/Configuration.mk
+++ b/Configuration.mk
@@ -176,7 +176,8 @@ override CPPFLAGS += \
       -fstack-usage\
       -D_FORTIFY_SOURCE=2\
       -Wall\
-      -Wextra
+      -Wextra\
+      -I$(TOCK_USERLAND_BASE_DIR)/newlib/include
 override WLFLAGS += \
       -Wl,--warn-common\
       -Wl,--gc-sections\
```

**Comment by bradjc** [comment]
I cannot get this version of newlib to build. I keep getting:

```
  CPPAS    libc/machine/arm/libc_a-setjmp.o
  CPPAS    libc/machine/arm/libc_a-strcmp.o
../../../../../../newlib-4.3.0.20230120/newlib/libc/machine/arm/setjmp.S: Assembler messages:
../../../../../../newlib-4.3.0.20230120/newlib/libc/machine/arm/setjmp.S:206: Error: selected processor does not support `vstm r0,{ d8-d15 }' in ARM mode
../../../../../../newlib-4.3.0.20230120/newlib/libc/machine/arm/setjmp.S:244: Error: selected processor does not support `vldm r0,{ d8-d15 }' in ARM mode
make[7]: *** [libc/machine/arm/libc_a-setjmp.o] Error 1
make[7]: *** Waiting for unfinished jobs....
  CC       libc/machine/arm/libc_a-strcpy.o
make[6]: *** [all] Error 2
make[5]: *** [multi-do] Error 1
make[4]: *** [all-multi] Error 2
make[3]: *** [all] Error 2
make[2]: *** [all-target-newlib] Error 2
make[1]: *** [all] Error 2
make: *** [rebuild-newlib] Error 2
```

I've tried gcc@13, gcc@11, the default gcc on mac, all with the same error.

Based on an earlier debug:

```
true "CC_FOR_BUILD=gcc-11" "CFLAGS=-g -Os -ffunction-sections -fdata-sections -fPIC -msingle-pic-base -mno-pic-data-is-text-relative  -marm -march=armv5te+fp -mfloat-abi=softfp" "CCASFLAGS=-g -Os -ffunction-sections -fdata-sections -fPIC -msingle-pic-base -mno-pic-data-is-text-relative  -marm -march=armv5te+fp -mfloat-abi=softfp" "CFLAGS_FOR_BUILD=-g -O2" "CFLAGS_FOR_TARGET=-g -Os -ffunction-sections -fdata-sections -fPIC -msingle-pic-base -mno-pic-data-is-text-relative" "INSTALL=/usr/local/bin/ginstall -c" "LDFLAGS=-marm -march=armv5te+fp -mfloat-abi=softfp" "LIBCFLAGS=-g -Os -ffunction-sections -fdata-sections -fPIC -msingle-pic-base -mno-pic-data-is-text-relative  -marm -march=armv5te+fp -mfloat-abi=softfp" "LIBCFLAGS_FOR_TARGET=-g -Os -ffunction-sections -fdata-sections -fPIC -msingle-pic-base -mno-pic-data-is-text-relative" "MAKE=/Library/Developer/CommandLineTools/usr/bin/make" "MAKEINFO=makeinfo --split-size=5000000 --split-size=5000000     " "PICFLAG=" "PICFLAG_FOR_TARGET=" "SHELL=/bin/sh" "EXPECT=expect" "RUNTEST=runtest" "RUNTESTFLAGS=" "exec_prefix=/usr/local" "infodir=/usr/local/share/info" "libdir=/usr/local/lib" "prefix=/usr/local" "tooldir=/usr/local/arm-none-eabi" "top_toollibdir=/usr/local/arm-none-eabi/lib/arm/v5te/softfp" "AR=arm-none-eabi-ar" "AS=arm-none-eabi-as" "CC=arm-none-eabi-gcc -B/Users/bradjc/git/libtock-c/newlib/newlib-arm-4.3.0.20230120-out/arm-none-eabi/arm/v5te/softfp/newlib/ -isystem /Users/bradjc/git/libtock-c/newlib/newlib-arm-4.3.0.20230120-out/arm-none-eabi/arm/v5te/softfp/newlib/targ-include -isystem /Users/bradjc/git/libtock-c/newlib/newlib-4.3.0.20230120/newlib/libc/include -B/Users/bradjc/git/libtock-c/newlib/newlib-arm-4.3.0.20230120-out/arm-none-eabi/arm/v5te/softfp/libgloss/arm -L/Users/bradjc/git/libtock-c/newlib/newlib-arm-4.3.0.20230120-out/arm-none-eabi/arm/v5te/softfp/libgloss/libnosys -L/Users/bradjc/git/libtock-c/newlib/newlib-4.3.0.20230120/libgloss/arm  -marm -march=armv5te+fp -mfloat-abi=softfp" "LD=arm-none-eabi-ld" "LIBCFLAGS=-g -Os -ffunction-sections -fdata-sections -fPIC -msingle-pic-base -mno-pic-data-is-text-relative  -marm -march=armv5te+fp -mfloat-abi=softfp" "NM=arm-none-eabi-nm" "PICFLAG=" "RANLIB=arm-none-eabi-ranlib" "DESTDIR=" DO=all multi-do # /Library/Developer/CommandLineTools/usr/bin/make
```

It seems like it is trying to build for `armv5te+fp`.

Did you run into any issues?

**Comment by bradjc** [comment]
I think it's time to:

1. Add the newlib header files to the repo.
2. Compile the c++ libraries for risc-v.
3. Remove all instructions to install newlib from the readme.

We've been cobbling this together for a long time and I think it's finally time to get this straighten out.

I'm working on compiling the C++ libraries for riscv.

**Comment by bradjc** [comment]
How did you build the cortex-m newlib 4.3.0 libraries? Which compiler version did you use?

I'm trying to build on Ubuntu 22.04 with arm-none-eabi-gcc 10.3.1 and I keep running into issues like:

```
  CC       libc/machine/arm/libc_a-aeabi_memset.o
  CPPAS    libc/machine/arm/libc_a-aeabi_memset-soft.o
  CC       libc/machine/arm/libc_a-aeabi_memclr.o
  CC       libc/machine/arm/libc_a-memchr-stub.o
../../../../../../newlib-4.3.0.20230120/newlib/libc/machine/arm/setjmp.S: Assembler messages:
../../../../../../newlib-4.3.0.20230120/newlib/libc/machine/arm/setjmp.S:206: Error: selected FPU does not support instruction -- `vstm r0,{ d8-d15 }'
../../../../../../newlib-4.3.0.20230120/newlib/libc/machine/arm/setjmp.S:244: Error: selected FPU does not support instruction -- `vldm r0,{ d8-d15 }'
  CPPAS    libc/machine/arm/libc_a-memchr.o
make[7]: *** [Makefile:16636: libc/machine/arm/libc_a-setjmp.o] Error 1
make[7]: *** Waiting for unfinished jobs....
  CC       libc/machine/arm/libc_a-memcpy-stub.o
  CPPAS    libc/machine/arm/libc_a-memcpy.o
make[7]: Leaving directory '/libtock-c/newlib/newlib-arm-4.3.0.20230120-out/arm-none-eabi/thumb/v7+fp/softfp/newlib'
```

**Comment by ppannuto** [comment]
> ../../../../../../newlib-4.3.0.20230120/newlib/libc/machine/arm/setjmp.S:206: Error: selected FPU does not support instruction -- `vstm r0,{ d8-d15 }'

That's a `FPv4-SP` instruction, i.e. part of the floating point extensions for the Cortex-M family; specifically, I think it's only available on Cortex-M4F and Cortex-M7 (and maybe Cortex-M32/33; don't know the newer ones off the top of my head as well). But that will definitely never work for any Cortex M0, and I don't think those extensions are an option for M3's.

That line looks like newlib is including some assembly file of its own writing, so it's not the complier emitting a bad instruction per se, but rather newlib choosing the wrong assembly stub file for the compilation?

**Comment by bradjc** [comment]
The offending line is here

https://github.com/bminor/newlib/blob/7236f22f713a923759ec7a98b4572c1910beb730/newlib/libc/machine/arm/setjmp.S#L227-L229

Which seems to be a relatively new addition. The `__ARM_FP` guard is maybe new to compilers too?

**Comment by alistair23** [comment]
I used just a standard ARM toolchain

```
arm-none-eabi-gcc (Arch Repository) 13.2.0
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

and I just checked today, no issues

**Comment by bradjc** [comment]
Yeah ok 4.3.0 compiles with gcc 13 but not with gcc 10.

I think we will need to separate this into two steps: 1) updating the build system and upgrading to 4.2.0, and 2) seeing what we can do about using 4.3.0. There are now two problems with 4.3.0: we can't build it on Ubuntu 22.04 (our CI) and versions compiled with gcc 13 don't seem to work for riscv on our CI.

**Comment by bradjc** [comment]
Implemented in #353

**Pull request closed without merging**